File: src/main/java/de/pflugradts/passbird/application/configuration/ConfigurationSyncService.java
package de.pflugradts.passbird.application.configuration;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.dataformat.yaml.YAMLFactory;
import com.google.inject.Inject;
import de.pflugradts.passbird.application.util.SystemOperation;
import lombok.AllArgsConstructor;
import lombok.NoArgsConstructor;

import static com.fasterxml.jackson.databind.MapperFeature.PROPAGATE_TRANSIENT_MARKER;
import static com.fasterxml.jackson.dataformat.yaml.YAMLGenerator.Feature.WRITE_DOC_START_MARKER;
import static de.pflugradts.passbird.application.configuration.ReadableConfiguration.CONFIGURATION_FILENAME;

@AllArgsConstructor
@NoArgsConstructor
public class ConfigurationSyncService implements ConfigurationSync {

    @Inject
    private UpdatableConfiguration updatableConfiguration;
    @Inject
    private SystemOperation systemOperation;

    @Override
    public void sync(final String directory) {
        updatableConfiguration.updateDirectory(directory);
        try {
            new ObjectMapper(new YAMLFactory()
                .disable(WRITE_DOC_START_MARKER))
                .enable(PROPAGATE_TRANSIENT_MARKER)
                .writeValue(
                    systemOperation.resolvePath(directory, CONFIGURATION_FILENAME).toFile(),
                    updatableConfiguration);
        } catch (Exception ex) {
            // FIXME error handling
        }
    }

}

File: src/main/kotlin/de/pflugradts/passbird/application/configuration/ConfigurationSyncService.kt
package de.pflugradts.passbird.application.configuration
import de.pflugradts.kotlinextensions.TryResult
import de.pflugradts.kotlinextensions.tryCatching
import de.pflugradts.passbird.application.Directory
import de.pflugradts.passbird.application.configuration.ReadableConfiguration.Companion.CONFIGURATION_FILENAME
import de.pflugradts.passbird.application.failure.ConfigurationFailure
import de.pflugradts.passbird.application.failure.reportFailure
import de.pflugradts.passbird.application.toFileName
import de.pflugradts.passbird.application.util.SystemOperation
class ConfigurationSyncService constructor(
    private val updatableConfiguration: UpdatableConfiguration,
    private val systemOperation: SystemOperation,
    private val configurationYamlMapper: ConfigurationYamlMapper = ConfigurationYamlMapper(),
) : ConfigurationSync {
    override fun sync(directory: Directory): TryResult<Unit> = writeConfiguration(directory) {
        updatableConfiguration.updateDirectory(directory)
    }

    override fun syncKeyStoreLocation(configurationDirectory: Directory, keyStoreDirectory: Directory): TryResult<Unit> =
        writeConfiguration(configurationDirectory) {
            updatableConfiguration.updateKeyStoreDirectory(keyStoreDirectory)
        }

    private fun writeConfiguration(directory: Directory, updateConfiguration: () -> Unit): TryResult<Unit> = tryCatching {
        updateConfiguration()
        val yaml = configurationYamlMapper.writeConfiguration(updatableConfiguration)
        systemOperation.writeToSensitiveFile(
            systemOperation.resolvePath(directory, CONFIGURATION_FILENAME.toFileName()),
        ) { outputStream ->
            outputStream.writer(Charsets.UTF_8).use { writer ->
                writer.write(yaml)
            }
        }
        Unit
    }.onFailure {
        reportFailure(ConfigurationFailure(it))
    }
}
